src/repeat-prevent.coffee

```diff
@@ -1,7 +1,39 @@
 # Description:
-#   A hubot script to do nothing
+#   A hubot script to prevent responses if they repeat a recent response
 #
 # Author:
 #   Andrew Lucas (sqweelygig) <andrewl@resin.io> <sqweelygig@gmail.com>
+moment = require 'moment'
+_ = require 'lodash'
+
+scopes = {}
+timeout = parseInt(process.env.HUBOT_PREVENT_REPEAT_TIMEOUT ? '30')
+
+# Remove old comments up every 10 percent of the way through the timeout
+maybeTidy = _.throttle(
+	->
+		horizon = moment().subtract(timeout, 'minutes')
+		for scope, comments of scopes
+			for comment, timestamp of comments when timestamp.isBefore horizon
+				delete scopes[scope][comment]
+	timeout * 6000 # (minutes->milliseconds * 10%)
+)
+
 module.exports = (robot) ->
-	return
+	robot.responseMiddleware (context, next, done) ->
+		# Extract and initialise the data we need
+		now = moment()
+		horizon = moment(now).subtract(timeout, 'minutes')
+		comment = JSON.stringify(context.strings)
+		scope = context.response.message.metadata?.thread_id ? context.response.message.room
+		scopes[scope] ?= {}
+
+		# If the comment isn't in our memory or is old
+		if (not scopes[scope][comment]?) or scopes[scope][comment].isBefore(horizon)
+			scopes[scope][comment] = now
+			next()
+		else
+			done()
+
+		# Trigger garbage collection
+		maybeTidy()
```

Comment:
Ah, I've just realised we might be overcomplicating this.

This only gets triggered on hubot responses, and only stores hubot's responses, is that right? In that case, that's actually not that often or that many, and the list of hubot responses you're looping over isn't that long either. In a half an hour expiry, we're probably talking about only 10s of messages tops.

Given that, you could stop throttling `tidy()` completely (since it's actually pretty cheap), do it explicitly at the start of this function, and then you don't have to use `horizon` or check `isBefore` afterwards at all, just in the `tidy` step.

The whole bot is then just: drop old messages, check for the new message is still in the list, and continue if not.

That assumes that indeed this is only on responses, and indeed we don't actually have _that_ many of those (every minute is fine, more than once a second probably isn't). Does that sound plausible?